### ISSUE TITLE
[rust] Unable to build

### ISSUE THREAD

**Issue by uberhacker** [skycocker/chromebrew#4510]
```
rust: Rust is a systems programming language that runs blazingly fast, prevents segfaults, and guarantees thread safety.
https://www.rust-lang.org/
version 1.47.0
No precompiled binary available for your platform, downloading source...
######################################################################## 100.0%
Archive downloaded
Unpacking archive using 'tar', this may take a while...
gcc (GCC) 10.2.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Building from source, this may take a while...
Rename all *.la files to *.la_tmp
Preconfiguring package...
info: downloading installer
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `4148885`,
 right: `0`', src/libstd/sys/unix/fs.rs:1225:25
stack backtrace:
   0: 0x773fdee4 - backtrace::backtrace::libunwind::trace::h585b9896cf7df742
                       at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: 0x773fdee4 - backtrace::backtrace::trace_unsynchronized::h15b719431e7282b4
                       at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: 0x773fdee4 - std::sys_common::backtrace::_print_fmt::h925402da86f620fd
                       at src/libstd/sys_common/backtrace.rs:78
   3: 0x773fdee4 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hcfe5f81dbb352f47
                       at src/libstd/sys_common/backtrace.rs:59
   4: 0x7706b480 - core::fmt::write::ha5eaf7a35a0eebb3
                       at src/libcore/fmt/mod.rs:1069
   5: 0x773fd76b - std::io::Write::write_fmt::h563accc0b115498b
                       at src/libstd/io/mod.rs:1504
   6: 0x773fd158 - std::sys_common::backtrace::_print::h3be4d6a3c9de807b
                       at src/libstd/sys_common/backtrace.rs:62
   7: 0x773fd158 - std::sys_common::backtrace::print::h6f16621e6cb34f44
                       at src/libstd/sys_common/backtrace.rs:49
   8: 0x773fd158 - std::panicking::default_hook::{{closure}}::hbaece5ec3201216c
                       at src/libstd/panicking.rs:198
   9: 0x773fca42 - std::panicking::default_hook::h73f9f65ea970381a
                       at src/libstd/panicking.rs:218
  10: 0x76ff3308 - rustup::currentprocess::with::{{closure}}::{{closure}}::h23130ad31300887a
  11: 0x773fc554 - std::panicking::rust_panic_with_hook::h586913f7baf1de1c
                       at src/libstd/panicking.rs:515
  12: 0x773fc32b - rust_begin_unwind
                       at src/libstd/panicking.rs:419
  13: 0x773fc2c5 - std::panicking::begin_panic_fmt::h32f7688c21deb376
                       at src/libstd/panicking.rs:373
  14: 0x77415c2a - std::sys::unix::fs::copy::h5b5168e8734e7171
                       at src/libstd/sys/unix/fs.rs:1225
  15: 0x773220e5 - rustup::utils::utils::copy_file::heb5ea6b427f79348
  16: 0x773db81f - rustup::cli::self_update::install_bins::h454e30a6942822c6
  17: 0x773de520 - rustup::cli::self_update::install::h4b794a1caf7857c7
  18: 0x773ea389 - rustup::cli::setup_mode::main::hd8a8c46e2dafda9c
  19: 0x76ff5fea - rustup_init::main::h75f2035d5a3186b1
  20: 0x77414872 - std::rt::lang_start_internal::{{closure}}::{{closure}}::h033c301fbc0c9761
                       at src/libstd/rt.rs:52
  21: 0x77414872 - std::sys_common::backtrace::__rust_begin_short_backtrace::hf6a0d326acb7358c
                       at src/libstd/sys_common/backtrace.rs:130
  22: 0x76ff6e93 - main
  23: 0x76dbf699 - __libc_start_main
rmdir: failed to remove ‘/usr/local/tmp’: Directory not empty
rust failed to build: `bash ./rustup.sh -y --no-modify-path --default-host i686-unknown-linux-gnu --default-toolchain stable --profile minimal` exited with 1
```

**Comment by satmandu** [comment]
```
chronos@m85x86_64 ~ $ crew install -s rust
rust: Rust is a systems programming language that runs blazingly fast, prevents segfaults, and guarantees thread safety.
https://www.rust-lang.org/
version 1.47.0
Downloading source...
############################################################################################ 100.0%                 -=O=-       #      #      #       #
Archive downloaded
Unpacking archive using 'tar', this may take a while...
gcc (GCC) 10.2.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Building from source, this may take a while...
Rename all *.la files to *.la_tmp
Preconfiguring package...
info: downloading installer
Warning: Not enforcing strong cipher suites for TLS, this is potentially less secure
info: profile set to 'minimal'
info: setting default host triple to x86_64-unknown-linux-gnu
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: latest update on 2020-10-08, rust version 1.47.0 (18bf6b4f0 2020-10-07)
info: downloading component 'cargo'
info: downloading component 'rust-std'
 21.2 MiB /  21.2 MiB (100 %)  13.0 MiB/s in  1s ETA:  0s
info: downloading component 'rustc'
 66.5 MiB /  66.5 MiB (100 %)  11.7 MiB/s in  6s ETA:  0s
info: installing component 'cargo'
info: Defaulting to 500.0 MiB unpack ram
info: installing component 'rust-std'
 21.2 MiB /  21.2 MiB (100 %)  11.2 MiB/s in  1s ETA:  0s
info: installing component 'rustc'
 66.5 MiB /  66.5 MiB (100 %)  12.8 MiB/s in  5s ETA:  0s
info: default toolchain set to 'stable'

  stable installed - rustc 1.47.0 (18bf6b4f0 2020-10-07)


Rust is installed now. Great!

To get started you need Cargo's bin directory
(/usr/local/tmp/crew/dest/usr/local/share/cargo/bin) in your PATH
environment variable.

To configure your current shell run
source /usr/local/tmp/crew/dest/usr/local/share/cargo/env
rmdir: failed to remove '/usr/local/tmp': Directory not empty
Rename all *.la_tmp files back to *.la
Performing pre-install...
Performing install...
tar: ./home/chronos: Cannot utime: Operation not permitted
tar: Exiting with failure status due to previous errors
Performing post-install...

To complete the installation, execute the following:
echo 'export CARGO_HOME=/usr/local/share/cargo' >> ~/.bashrc
echo 'export RUSTUP_HOME=/usr/local/share/rustup' >> ~/.bashrc
source ~/.bashrc

To add bash completion for cargo, execute the following:
echo '# cargo completion' >> ~/.bashrc
echo 'if [ -f /usr/local/share/bash-completion/completions/cargo ]; then' >> ~/.bashrc
echo '  source /usr/local/share/bash-completion/completions/cargo' >> ~/.bashrc
echo 'fi' >> ~/.bashrc
source ~/.bashrc

To add bash completion for rustup, execute the following:
echo '# rustup completion' >> ~/.bashrc
echo 'if [ -f /usr/local/share/bash-completion/completions/rustup ]; then' >> ~/.bashrc
echo '  source /usr/local/share/bash-completion/completions/rustup' >> ~/.bashrc
echo 'fi' >> ~/.bashrc
source ~/.bashrc

Rust installed!
chronos@m85x86_64 ~ $ 
```

Maybe try ```ENV['RUSTUP_IO_THREADS'] = '1'``` as per https://stackoverflow.com/questions/61603982/thread-main-panicked-at-assertion-failed-left-right-left-22-right ?

**Comment by satmandu** [comment]
works for me on i686:
```
chronos@m58i686 /usr/local/tmp $ crew install -s rust
rust: Rust is a systems programming language that runs blazingly fast, prevents segfaults, and guarantees thread safety.
https://www.rust-lang.org/
version 1.47.0
Downloading source...
############################################################################################################### 100.0%                    -=O=-                    #        #        #        #                                            
Archive downloaded
Unpacking archive using 'tar', this may take a while...
gcc (GCC) 10.2.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Building from source, this may take a while...
Rename all *.la files to *.la_tmp
Preconfiguring package...
info: downloading installer
Warning: Not enforcing strong cipher suites for TLS, this is potentially less secure
info: profile set to 'minimal'
info: setting default host triple to i686-unknown-linux-gnu
info: syncing channel updates for 'stable-i686-unknown-linux-gnu'
info: latest update on 2020-10-08, rust version 1.47.0 (18bf6b4f0 2020-10-07)
info: downloading component 'cargo'
info: downloading component 'rust-std'
 19.9 MiB /  19.9 MiB (100 %)  13.2 MiB/s in  1s ETA:  0s
info: downloading component 'rustc'
 69.1 MiB /  69.1 MiB (100 %)  13.3 MiB/s in  5s ETA:  0s
info: installing component 'cargo'
info: Defaulting to 500.0 MiB unpack ram
info: installing component 'rust-std'
 19.9 MiB /  19.9 MiB (100 %)  10.5 MiB/s in  1s ETA:  0s
info: installing component 'rustc'
 69.1 MiB /  69.1 MiB (100 %)  12.0 MiB/s in  5s ETA:  0s
info: default toolchain set to 'stable'

  stable installed - rustc 1.47.0 (18bf6b4f0 2020-10-07)


Rust is installed now. Great!

To get started you need Cargo's bin directory 
(/usr/local/tmp/crew/dest/usr/local/share/cargo/bin) in your PATH
environment variable.

To configure your current shell run 
source /usr/local/tmp/crew/dest/usr/local/share/cargo/env
rmdir: failed to remove '/usr/local/tmp': Directory not empty
Rename all *.la_tmp files back to *.la
Performing pre-install...
Performing install...
tar: ./home/chronos: Cannot utime: Operation not permitted
tar: Exiting with failure status due to previous errors
Performing post-install...

To complete the installation, execute the following:
echo 'export CARGO_HOME=/usr/local/share/cargo' >> ~/.bashrc
echo 'export RUSTUP_HOME=/usr/local/share/rustup' >> ~/.bashrc
source ~/.bashrc

To add bash completion for cargo, execute the following:
echo '# cargo completion' >> ~/.bashrc
echo 'if [ -f /usr/local/share/bash-completion/completions/cargo ]; then' >> ~/.bashrc
echo '  source /usr/local/share/bash-completion/completions/cargo' >> ~/.bashrc
echo 'fi' >> ~/.bashrc
source ~/.bashrc

To add bash completion for rustup, execute the following:
echo '# rustup completion' >> ~/.bashrc
echo 'if [ -f /usr/local/share/bash-completion/completions/rustup ]; then' >> ~/.bashrc
echo '  source /usr/local/share/bash-completion/completions/rustup' >> ~/.bashrc
echo 'fi' >> ~/.bashrc
source ~/.bashrc

Rust installed!
chronos@m58i686 /usr/local/tmp $ 
```

**Comment by uberhacker** [comment]
It installs but does it build?

**Comment by satmandu** [comment]
```
crew build rust
rust: Rust is a systems programming language that runs blazingly fast, prevents segfaults, and guarantees thread safety.
https://www.rust-lang.org/
version 1.47.0
Downloading source...
############################################################################################################### 100.0%                  -=O=-                      #        #        #        #                                            
Archive downloaded
Unpacking archive using 'tar', this may take a while...
Building from source, this may take a while...
Rename all *.la files to *.la_tmp
Preconfiguring package...
info: downloading installer
Warning: Not enforcing strong cipher suites for TLS, this is potentially less secure
warning: it looks like you have an existing installation of Rust at:
warning: /usr/local/bin
warning: rustup should not be installed alongside Rust. Please uninstall your existing Rust first.
warning: Otherwise you may have confusion unless you are careful with your PATH
warning: If you are sure that you want both rustup and your already installed Rust
warning: then please reply `y' or `yes' or set RUSTUP_INIT_SKIP_PATH_CHECK to yes
warning: or pass `-y' to ignore all ignorable checks.
error: cannot install while Rust is installed
warning: continuing (because the -y flag is set and the error is ignorable)
info: profile set to 'minimal'
info: setting default host triple to i686-unknown-linux-gnu
info: syncing channel updates for 'stable-i686-unknown-linux-gnu'
info: latest update on 2020-10-08, rust version 1.47.0 (18bf6b4f0 2020-10-07)
info: downloading component 'cargo'
info: downloading component 'rust-std'
 19.9 MiB /  19.9 MiB (100 %)  14.0 MiB/s in  1s ETA:  0s
info: downloading component 'rustc'
 69.1 MiB /  69.1 MiB (100 %)  13.8 MiB/s in  5s ETA:  0s
info: installing component 'cargo'
info: Defaulting to 500.0 MiB unpack ram
info: installing component 'rust-std'
 19.9 MiB /  19.9 MiB (100 %)   7.6 MiB/s in  2s ETA:  0s
info: installing component 'rustc'
 69.1 MiB /  69.1 MiB (100 %)   8.2 MiB/s in  8s ETA:  0s
info: default toolchain set to 'stable'

  stable installed - rustc 1.47.0 (18bf6b4f0 2020-10-07)


Rust is installed now. Great!

To get started you need Cargo's bin directory 
(/usr/local/tmp/crew/dest/usr/local/share/cargo/bin) in your PATH
environment variable.

To configure your current shell run 
source /usr/local/tmp/crew/dest/usr/local/share/cargo/env
rmdir: failed to remove '/usr/local/tmp': Directory not empty
Rename all *.la_tmp files back to *.la
Checking...
Archiving...
rust is built!
```

**Comment by uberhacker** [comment]
I really want to give up supporting i686.  It's just too much maintenance at this point.

**Comment by satmandu** [comment]
I'm not an i686 users, but the packages are building perfectly for me in containers based upon two separate i686 recovery images. Are you building these packages natively on the hardware? That sounds insane if you don't have to...

**Comment by uberhacker** [comment]
I checked out your chromeos_docker.sh script and it looks useful!  The problem with docker is it doesn't accurately simulate actual Chromebook hardware.  Your script also adds paths to `LD_LIBRARY_PATH`.  This is not what a Chromebook environment looks like out of the box.  We can certainly do all kinds of crazy configs to make it work but that will most certainly not provide a suitable testing environment.  We should concentrate on taking a minimalist approach to packaging.  That means no extra config and cloning the repo from master fresh to provide a clean testing environment in docker.

**Comment by satmandu** [comment]
Agreed on the LD_LIBRARY_PATH hackery being, well, a hack. 

I was just happy to get a container working and installed. (Python needing to be installed twice to work was something I wasn't expecting.)

Next, I'm trying with the LD_LIBRARY_PATH stuff taken out.

**Comment by satmandu** [comment]
Just took out all the LD_LIBRARY_PATH setting lines, and I'm still not seeing any issues compiling or building, even with the alex recovery image/

Is that weird?

My big difference is that I am starting off with installing gcc10 in lieu of gcc8.

**Comment by uberhacker** [comment]
Fixed in #4552.
